src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/ClientInterface.cs

```diff
@@ -0,0 +1,453 @@
+        //
+        static ClientInterface()
+        {
+            // The max length must not be negative.
+            //
+            // C# doesn't have compile-time type traits, so we can't confirm
+            // that MAXLEN_CLIENTINTERFACE is unsigned.  Instead, we capture
+            // it into a ushort, and let the compiler decide if that is a
+            // permitted operation.
+            //
+            ushort maxLen = TdsEnums.MAXLEN_CLIENTINTERFACE;
+
+            // Determine the OS type.
+            //
+            // This is done outside of Build() to allow tests to inject
+            // specific values.
+            //
+            string osType = Unknown;
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                osType = "Windows";
+            }
+            else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+            {
+                osType = "Linux";
+            }
+            else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+            {
+                osType = "macOS";
+            }
+// The FreeBSD platform doesn't exist in .NET Framework at all.
+#if NET
+            else if (RuntimeInformation.IsOSPlatform(OSPlatform.FreeBSD))
+            {
+                osType = "FreeBSD";
+            }
+#endif // NET
+
+            // Build it!
+            _name = Build(
+                maxLen,
+                DriverName,
+                osType,
+                RuntimeInformation.ProcessArchitecture,
+                RuntimeInformation.OSDescription,
+                RuntimeInformation.FrameworkDescription);
+        }
+
+        // Build the Client Interface Name and return it.
+        //
+        // The length of the returned value will never be longer than maxLen.
+        //
+        // All known exceptions are consumed.
+        //
+        public static string Build(
+            ushort maxLen,
+            string driverName,
+            string osType,
+            Architecture arch,
+            string osInfo,
+            string runtimeInfo)
+        {
+            string result;
+
+            // Clean and truncate the driver name, max length 10.  We will need
+            // it for error handling.
+            driverName = Trunc(Clean(driverName), 10);
+
+            try
+            {
+                // Expect to build a string whose length is up to our max
+                // length.
+                //
+                // This isn't a max capacity, but a hint for initial buffer
+                // allocation.  We will truncate to our max length after all of
+                // the pieces have been appended.
+                //
+                StringBuilder name = new StringBuilder(maxLen);
+
+                // Start with the (clean) driver name.
+                name.Append(driverName);
+                name.Append('|');
+
+                // Add the OS Type, max length 10.
+                name.Append(Trunc(Clean(osType), 10));
```

Comment:
Magic number 10 should be extracted into a const or readonly int. OSTYPE_MAX_LEN or something on these lines.